src/circuit/auto-relay.ts

```diff
@@ -0,0 +1,284 @@
+import { logger } from '@libp2p/logger'
+import { fromString as uint8ArrayFromString } from 'uint8arrays/from-string'
+import { toString as uint8ArrayToString } from 'uint8arrays/to-string'
+import { RELAY_CODEC } from './multicodec.js'
+import { canHop } from './circuit/hop.js'
+import { namespaceToCid } from './utils.js'
+import {
+  CIRCUIT_PROTO_CODE,
+  HOP_METADATA_KEY,
+  HOP_METADATA_VALUE,
+  RELAY_RENDEZVOUS_NS
+} from './constants.js'
+import type { PeerId } from '@libp2p/interfaces/peer-id'
+import type { AddressSorter, PeerProtocolsChangeData } from '@libp2p/interfaces/peer-store'
+import type { Connection } from '@libp2p/interfaces/connection'
+import type { Components } from '@libp2p/interfaces/components'
+import sort from 'it-sort'
+import all from 'it-all'
+import { pipe } from 'it-pipe'
+import { publicAddressesFirst } from '@libp2p/utils/address-sort'
+
+const log = logger('libp2p:auto-relay')
+
+const noop = () => {}
+
+export interface AutoRelayInit {
+  addressSorter?: AddressSorter
+  maxListeners?: number
+  onError?: (error: Error, msg?: string) => void
+}
+
+export class AutoRelay {
+  private readonly components: Components
+  private readonly addressSorter: AddressSorter
+  private readonly maxListeners: number
+  private readonly listenRelays: Set<string>
+  private readonly onError: (error: Error, msg?: string) => void
+
+  constructor (components: Components, init: AutoRelayInit) {
+    this.components = components
+    this.addressSorter = init.addressSorter ?? publicAddressesFirst
+    this.maxListeners = init.maxListeners ?? 1
+    this.listenRelays = new Set()
+    this.onError = init.onError ?? noop
+
+    this._onProtocolChange = this._onProtocolChange.bind(this)
+    this._onPeerDisconnected = this._onPeerDisconnected.bind(this)
+
+    this.components.getPeerStore().addEventListener('change:protocols', (evt) => {
+      void this._onProtocolChange(evt).catch(err => {
+        log.error(err)
+      })
+    })
+    this.components.getConnectionManager().addEventListener('peer:disconnect', this._onPeerDisconnected)
+  }
+
+  /**
+   * Check if a peer supports the relay protocol.
+   * If the protocol is not supported, check if it was supported before and remove it as a listen relay.
+   * If the protocol is supported, check if the peer supports **HOP** and add it as a listener if
+   * inside the threshold.
+   */
+  async _onProtocolChange (evt: CustomEvent<PeerProtocolsChangeData>) {
```

Comment:
We don't use private because of tests or is it too much effort to change them to private?

Reply:
The external API for the components are largely defined by interfaces, but it's useful to be able to call internal methods in the tests, but yes, this could change to be private.

I tried out [hash-prefixed private](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) in the [EventEmitter](https://github.com/libp2p/js-libp2p-interfaces/blob/master/packages/libp2p-interfaces/src/index.ts#L77) class - it works but causes a warning to be logged to the console during the tests, kind of want to experiment with that a bit more before rolling it out everywhere.

In the interim my hope is the interface-as-public api should stop people calling private(ish) methods.